Core/include/Acts/EventData/Charge.hpp

```diff
@@ -0,0 +1,149 @@
+/// constructed with the input charge magnitude
+///
+/// ```cpp
+/// Charge c(1_e);
+/// ```
+///
+/// and can then be used to extract the charge value
+///
+/// ```cpp
+/// auto q = c.extractCharge(qOverP);
+/// ```
+///
+/// or the absolute momentum
+///
+/// ```cpp
+/// auto p = c.extractMomentum(qOverP);
+/// ```
+///
+/// from the charge-over-momentum track parameter.
+///
+/// @{
+
+/// Charge and momentum interpretation for neutral particles.
+struct Neutral {
+  Neutral() = default;
+  /// Construct and verify the input charge magnitude (in debug builds).
+  ///
+  /// This constructor is only provided to allow consistent construction.
+  template <typename T>
+  constexpr Neutral(T absQ) noexcept {
+    assert((absQ == 0.0f) and "Input charge must be zero");
+    // suppress `unused variable` warning in non-debug builds
+    (void)(absQ);
+  }
+
+  template <typename T>
+  constexpr T extractCharge(T /* pInv */) const noexcept {
+    return 0;
+  }
+  template <typename T>
+  constexpr T extractMomentum(T pInv) const noexcept {
+    // the abs is not strictly needed. it is added to protect against invalid,
+    // i.e. negative, 1/p values to ensure that the output is still correct.
+    return std::abs(1 / pInv);
+  }
+
+  /// Compare for equality.
+  ///
+  /// This is always `true` as `Neutral` has no internal state.
+  /// Must be available to provide a consistent interface.
+  friend constexpr bool operator==(Neutral, Neutral) noexcept { return true; }
+};
+
+/// Charge and momentum interpretation for particles with +-e charge.
+struct SinglyCharged {
+  SinglyCharged() = default;
+  /// Construct and verify the input charge magnitude (in debug builds).
+  ///
+  /// This constructor is only provided to allow consistent construction.
+  template <typename T>
+  constexpr SinglyCharged(T absQ) noexcept {
+    assert((absQ == static_cast<T>(UnitConstants::e)) and
+           "Input charge magnitude must be e");
+    // suppress `unused variable` warning in non-debug builds
+    (void)(absQ);
+  }
+
+  template <typename T>
+  constexpr T extractCharge(T qOverP) const noexcept {
+    return std::copysign(static_cast<T>(UnitConstants::e), qOverP);
+  }
+  template <typename T>
+  constexpr T extractMomentum(T qOverP) const noexcept {
+    return std::abs(static_cast<T>(UnitConstants::e) / qOverP);
+  }
+
+  /// Compare for equality.
+  ///
+  /// This is always `true` as `SinglyCharged` has no internal state.
+  /// Must be available to provide a consistent interface.
+  friend constexpr bool operator==(SinglyCharged, SinglyCharged) noexcept {
+    return true;
+  }
+};
+
+/// Charge and momentum interpretation for arbitrarily charged particles.
+///
+/// Only a charge magnitude identical to zero is interpreted as representing a
+/// neutral particle. This avoids ambiguities that might arise from using an
+/// approximate comparison with an arbitrary epsilon.
+struct AnyCharge {
+  float magnitude;
```

Comment:
I understand the choice of float here, but in some sense we should try to get this harmonised.

Reply:
One could template this on the underlying float type such that this is always consistent with the scalar type of the corresponding parameters vector. I am not sure if we gain anything, except a whole lot of nested templates that are hard to read.

The main use case for this are simulations where we do not want to treat neutral and charge particles separately and heavy ion reconstruction where we might want to reconstruct multiple charged particles. The charge will be integer values for both cases, but I did not want to restrict it to integers on the off-chance that somebody is doing strange things, e.g. fractional charge new physics particles. In all these cases, `float` is fully sufficient to store all possible values without loss of precision and making this configurable would only waste storage.

Reply:
The magnitude is now a private member variable.